addons/product/models/product_attribute.py

```diff
@@ -38,18 +45,32 @@ def write(self, vals):
         This is important to prevent because changing the type would make
         existing combinations invalid without recomputing them, and recomputing
         them might take too long and we don't want to change products without
-        the user knowing about it."""
+        the user knowing about it.
+
+        We also need to invalidate the cache when changing the sequence because
+        the prefeteched o2m `attribute_line_ids` has to be resequenced.
+        """
         if 'create_variant' in vals:
             products = self._get_related_product_templates()
             if products:
                 message = ', '.join(products.mapped('name'))
                 raise UserError(_('You are trying to change the type of an attribute value still referenced on at least one product template: %s') % message)
-        invalidate_cache = 'sequence' in vals and any(record.sequence != vals['sequence'] for record in self)
+        to_invalidate = self.filtered(lambda pa: pa.sequence != vals['sequence']) if 'sequence' in vals else self.env['product.attribute']
+
         res = super(ProductAttribute, self).write(vals)
-        if invalidate_cache:
-            # prefetched o2m have to be resequenced
-            # (eg. product.template: attribute_line_ids)
-            self.invalidate_cache()
+
+        if to_invalidate:
+            self.env['product.attribute'].invalidate_cache(fnames=['attribute_line_ids'], ids=to_invalidate.ids)
+            self.env['product.template'].invalidate_cache(fnames=[
+                'attribute_line_ids',
```

Comment:
if we keep this, we should probably add this list in a constant of product.template, since it is already duplicated in this commit.

but are we sure that we want all this code instead of just invalidate a little bit more that what is needed?

//cc @rco-odoo

Reply:
This is a good question, I'm not sure if it's worth computing which ids have to be invalidated like in the PR when it is more complicated than a simple map.

But at least specifying which model and which fields to clear is for sure better for performance, and probably more readable too (especially if following your suggestion to store the list in a variable).